src/v/cluster/controller.cc

```diff
@@ -548,59 +574,45 @@ ss::future<> controller::create_cluster() {
  * to it, and before we have started communicating with peers.
  */
 ss::future<> controller::cluster_creation_hook(cluster_discovery& discovery) {
-    auto is_cluster_founder = co_await discovery.is_cluster_founder();
-    if (!is_cluster_founder) {
-        co_return;
-    }
+    const bool controller_log_exists = _raft0->last_visible_index()
+                                       > model::offset{};
 
-    if (config::node().empty_seed_starts_cluster()) {
-        // This is the legacy empty-seed-starts-cluster mode
+    if (co_await discovery.is_cluster_founder()) {
         vassert(
-          config::node().seed_servers().empty(), "We are not on the root node");
-
-        if (
-          _raft0->last_visible_index() > model::offset{}
-          || _raft0->config().brokers().size() > 1) {
-            // The controller log has already been written to
-            co_return;
+          !controller_log_exists,
+          "Can't be a cluster founder with controller log present");
+        if (config::node().empty_seed_starts_cluster()) {
+            vassert(
+              config::node().seed_servers().empty(),
+              "Cluster founder is not the root node");
+            vassert(
+              _raft0->config().brokers().size() == 1,
+              "Root is a cluster founder but joiners are in the cluster alrdy");
```

Comment:
nit: already

Reply:
that's intentional to avoid carryover to next line. I think that's ok for vassert :)